RNTuple endian issues

- [x] Checked for duplicates

### Describe the bug

Test failures in RNTuple on s390x. Endian issues.
There are way fewer failures in 6.28.00 than in earlier versions, this report is about remaining ones.

### Expected behavior

RNTuple should work for big-endian too.

### To Reproduce

1. Build root 6.28.00 for s390x
2. Run unit tests

### Setup

1. ROOT version: 6.28.00
2. Operating system: Fedora Linux on s390x, RHEL+EPEL on s390x
3. Compilation from source.

### Additional context

The failing tests 

#### gtest-tree-ntuple-v7-test-ntuple-basics

Fatal: nread == nbytes violated
~~~
 474/1272 Test  #421: gtest-tree-ntuple-v7-test-ntuple-basics .............................***Failed   63.64 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 17 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 13 tests from RNTuple
[ RUN      ] RNTuple.ReconstructModel
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.ReconstructModel (160 ms)
[ RUN      ] RNTuple.MultipleInFile
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Fatal: nread == nbytes violated at line 1083 of `/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RMiniFile.cxx'
aborting
[ --- ]
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: Subprocess aborted
~~~

#### gtest-tree-ntuple-v7-test-ntuple-endian

Here the bytes are in the right order, but shifted 4 bytes.

$ printf '%016x\n%016x\n' 1234605615003729920 287454020
1122334400000000
0000000011223344
$ printf '%016x\n%016x\n' 6153737371142586367 1432778632
55667788ffffffff
0000000055667788
~~~
 449/1272 Test  #424: gtest-tree-ntuple-v7-test-ntuple-endian .............................***Failed    0.04 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 9 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 9 tests from RColumnElementEndian
[ RUN      ] RColumnElementEndian.Float
[       OK ] RColumnElementEndian.Float (6 ms)
[ RUN      ] RColumnElementEndian.Double
[       OK ] RColumnElementEndian.Double (3 ms)
[ RUN      ] RColumnElementEndian.Int16
[       OK ] RColumnElementEndian.Int16 (0 ms)
[ RUN      ] RColumnElementEndian.UInt16
[       OK ] RColumnElementEndian.UInt16 (0 ms)
[ RUN      ] RColumnElementEndian.Int32
[       OK ] RColumnElementEndian.Int32 (0 ms)
[ RUN      ] RColumnElementEndian.UInt32
[       OK ] RColumnElementEndian.UInt32 (0 ms)
[ RUN      ] RColumnElementEndian.Int64
[       OK ] RColumnElementEndian.Int64 (0 ms)
[ RUN      ] RColumnElementEndian.UInt64
[       OK ] RColumnElementEndian.UInt64 (0 ms)
[ RUN      ] RColumnElementEndian.Int64Int32
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_endian.cxx:286: Failure
Expected equality of these values:
  ints[0]
    Which is: 1234605615003729920
  0x0000000011223344
    Which is: 287454020
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_endian.cxx:287: Failure
Expected equality of these values:
  ints[1]
    Which is: 6153737371142586367
  0x0000000055667788
    Which is: 1432778632
[  FAILED  ] RColumnElementEndian.Int64Int32 (0 ms)
[----------] 9 tests from RColumnElementEndian (13 ms total)
[----------] Global test environment tear-down
[==========] 9 tests from 1 test suite ran. (13 ms total)
[  PASSED  ] 8 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] RColumnElementEndian.Int64Int32
 1 FAILED TEST
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 1
~~~

#### gtest-tree-ntuple-v7-test-ntuple-extended

This looks like an endian error:

$ printf '%x\n' 1677721600
64000000

Header length shoud be 0x00000064 = 100
~~~
 501/1272 Test  #443: gtest-tree-ntuple-v7-test-ntuple-extended ...........................***Failed   81.93 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from RNTuple
[ RUN      ] RNTuple.RealWorld1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.RealWorld1 (5260 ms)
[ RUN      ] RNTuple.RandomAccess
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.RandomAccess (229 ms)
[ RUN      ] RNTuple.LargeFile1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
Error in <TFile::Init>: file test_large_file1.root has an incorrect header length (1677721600) or incorrect end of file length (-1760683648)
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_extended.cxx:159: Failure
Value of: f
  Actual: false
Expected: true
 *** Break *** segmentation violation
[ --- ]
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 139
~~~

#### gtest-tree-ntuple-v7-test-ntuple-minifile

$ printf '%x\n' 1677721600
64000000

Header length shoud be 0x00000064 = 100

Fatal: nread == nbytes violated
~~~
 478/1272 Test  #440: gtest-tree-ntuple-v7-test-ntuple-minifile ...........................***Failed   65.92 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 8 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 8 tests from MiniFile
[ RUN      ] MiniFile.Raw
[       OK ] MiniFile.Raw (0 ms)
[ RUN      ] MiniFile.Stream
Error in <TFile::ReadBuffer>: error reading all requested bytes from file test_ntuple_minifile_stream.root, got 0 of 6816316
Error in <TFile::Init>: test_ntuple_minifile_stream.root failed to read the file header information at 1677721600 (size=6816316)
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_minifile.cxx:84: Failure
Value of: file
  Actual: false
Expected: true
[  FAILED  ] MiniFile.Stream (191 ms)
[ RUN      ] MiniFile.Proper
Fatal: nread == nbytes violated at line 1083 of `/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RMiniFile.cxx'
aborting
[ --- ]
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: Subprocess aborted
~~~

#### gtest-tree-ntuple-v7-test-ntuple-rdf

$ printf '%x\n' 1677721600
64000000

Header length shoud be 0x00000064 = 100
~~~
 491/1272 Test  #431: gtest-tree-ntuple-v7-test-ntuple-rdf ................................***Failed   74.39 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from RNTuple
[ RUN      ] RNTuple.RDF
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Error in <TFile::Init>: file test_ntuple_rdf.root has an incorrect header length (1677721600) or incorrect end of file length (-4176807177076801536)
 *** Break *** segmentation violation
[ --- ]
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 139
~~~
#### gtest-tree-ntuple-v7-test-ntuple-serialize

Endian issue: 256 = 0x0100, expected 0x0001 = 1.
~~~
 457/1272 Test  #432: gtest-tree-ntuple-v7-test-ntuple-serialize ..........................***Failed    0.24 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 14 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 14 tests from RNTuple
[ RUN      ] RNTuple.SerializeInt
[       OK ] RNTuple.SerializeInt (0 ms)
[ RUN      ] RNTuple.SerializeString
[       OK ] RNTuple.SerializeString (0 ms)
[ RUN      ] RNTuple.SerializeColumnType
[       OK ] RNTuple.SerializeColumnType (0 ms)
[ RUN      ] RNTuple.SerializeFieldStructure
[       OK ] RNTuple.SerializeFieldStructure (0 ms)
[ RUN      ] RNTuple.SerializeEnvelope
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_serialize.cxx:121: Failure
Value of: err.what()
Expected: has substring "CRC32"
  Actual: 0x2aa065e7348 pointing to "The RNTuple format is too new (version 256) (unchecked RResult access!)\nAt:\n  static RResult<unsigned int> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeEnvelope(const void*, uint32_t, uint32_t&) [/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:629]\n  static RResult<unsigned int> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeEnvelope(const void*, uint32_t) [/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:646]\n" (of type char const*)
[  FAILED  ] RNTuple.SerializeEnvelope (0 ms)
[ RUN      ] RNTuple.SerializeFrame
[       OK ] RNTuple.SerializeFrame (0 ms)
[ RUN      ] RNTuple.SerializeFeatureFlags
[       OK ] RNTuple.SerializeFeatureFlags (0 ms)
[ RUN      ] RNTuple.SerializeLocator
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_serialize.cxx:341: Failure
Failed
unsupported locator type should throw
[  FAILED  ] RNTuple.SerializeLocator (0 ms)
[ RUN      ] RNTuple.SerializeEnvelopeLink
[       OK ] RNTuple.SerializeEnvelopeLink (0 ms)
[ RUN      ] RNTuple.SerializeClusterSummary
[       OK ] RNTuple.SerializeClusterSummary (0 ms)
[ RUN      ] RNTuple.SerializeClusterGroup
[       OK ] RNTuple.SerializeClusterGroup (0 ms)
[ RUN      ] RNTuple.SerializeEmptyHeader
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.SerializeEmptyHeader (190 ms)
[ RUN      ] RNTuple.SerializeHeader
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.SerializeHeader (0 ms)
[ RUN      ] RNTuple.SerializeFooter
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.SerializeFooter (0 ms)
[----------] 14 tests from RNTuple (190 ms total)
[----------] Global test environment tear-down
[==========] 14 tests from 1 test suite ran. (190 ms total)
[  PASSED  ] 12 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] RNTuple.SerializeEnvelope
[  FAILED  ] RNTuple.SerializeLocator
 2 FAILED TESTS
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 1
~~~

#### gtest-tree-ntuple-v7-test-ntuple-types

Endian issue:

$ printf '%x\n%x\n' 180388626432 42
2a00000000
2a
~~~
 465/1272 Test  #433: gtest-tree-ntuple-v7-test-ntuple-types ..............................***Failed    5.68 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 18 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 18 tests from RNTuple
[ RUN      ] RNTuple.TypeName
[       OK ] RNTuple.TypeName (316 ms)
[ RUN      ] RNTuple.CreateField
[       OK ] RNTuple.CreateField (0 ms)
[ RUN      ] RNTuple.ArrayField
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.ArrayField (23 ms)
[ RUN      ] RNTuple.StdPair
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.StdPair (226 ms)
[ RUN      ] RNTuple.StdTuple
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.StdTuple (169 ms)
[ RUN      ] RNTuple.Int64_t
[       OK ] RNTuple.Int64_t (0 ms)
[ RUN      ] RNTuple.Char
[       OK ] RNTuple.Char (0 ms)
[ RUN      ] RNTuple.Int8_t
[       OK ] RNTuple.Int8_t (0 ms)
[ RUN      ] RNTuple.Int16_t
[       OK ] RNTuple.Int16_t (0 ms)
[ RUN      ] RNTuple.UInt16_t
[       OK ] RNTuple.UInt16_t (0 ms)
[ RUN      ] RNTuple.UnsupportedStdTypes
[       OK ] RNTuple.UnsupportedStdTypes (3 ms)
[ RUN      ] RNTuple.Casting
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_types.cxx:280: Failure
Expected equality of these values:
  42
  *fieldCast
    Which is: 180388626432
[  FAILED  ] RNTuple.Casting (56 ms)
[ RUN      ] RNTuple.TClass
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.TClass (284 ms)
[ RUN      ] RNTuple.TClassMultipleInheritance
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.TClassMultipleInheritance (195 ms)
[ RUN      ] RNTuple.TClassTemplatedBase
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.TClassTemplatedBase (260 ms)
[ RUN      ] RNTuple.TVirtualCollectionProxy
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.TVirtualCollectionProxy (4036 ms)
[ RUN      ] RNTuple.Enums
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.Enums (29 ms)
[ RUN      ] RNTuple.Traits
[       OK ] RNTuple.Traits (0 ms)
[----------] 18 tests from RNTuple (5604 ms total)
[----------] Global test environment tear-down
[==========] 18 tests from 1 test suite ran. (5604 ms total)
[  PASSED  ] 17 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] RNTuple.Casting
 1 FAILED TEST
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: 1
~~~

--- a/tree/ntuple/v7/inc/ROOT/RColumnElement.hxx
+++ b/tree/ntuple/v7/inc/ROOT/RColumnElement.hxx
@@ -641,8 +641,9 @@ public:
       std::uint16_t *uint16Array = reinterpret_cast<std::uint16_t *>(src);
 
       for (std::size_t i = 0; i < count; ++i) {
-         ByteSwapIfNecessary(floatArray[i]);
-         floatArray[i] = HalfToFloat(uint16Array[i]);
+         std::uint16_t val = uint16Array[i];
+         ByteSwapIfNecessary(val);
+         floatArray[i] = HalfToFloat(val);
       }
    }
 };

--- a/tree/ntuple/v7/inc/ROOT/RColumnElement.hxx
+++ b/tree/ntuple/v7/inc/ROOT/RColumnElement.hxx
@@ -580,7 +580,7 @@ public:
          RSwitchElement element{srcArray[i].GetIndex(), srcArray[i].GetTag()};
 #if R__LITTLE_ENDIAN == 0
          element.fIndex = RByteSwap<8>::bswap(element.fIndex);
-         element.fTag = RByteSwap<8>::bswap(element.fTag);
+         element.fTag = RByteSwap<4>::bswap(element.fTag);
 #endif
          memcpy(dstArray + i * 12, &element, 12);
       }
@@ -595,7 +595,7 @@ public:
          memcpy(&element, srcArray + i * 12, 12);
 #if R__LITTLE_ENDIAN == 0
          element.fIndex = RByteSwap<8>::bswap(element.fIndex);
-         element.fTag = RByteSwap<8>::bswap(element.fTag);
+         element.fTag = RByteSwap<4>::bswap(element.fTag);
 #endif
          dstArray[i] = ROOT::Experimental::RColumnSwitch(ClusterSize_t{element.fIndex}, element.fTag);
       }

--- a/tree/ntuple/v7/src/RMiniFile.cxx
+++ b/tree/ntuple/v7/src/RMiniFile.cxx
@@ -56,7 +56,7 @@ private:
    static std::uint16_t Swap(std::uint16_t val)
    {
 #if R__LITTLE_ENDIAN == 1
-      return RByteSwap<sizeof(fValBE)>::bswap(val);
+      return RByteSwap<sizeof(val)>::bswap(val);
 #else
       return val;
 #endif
@@ -80,7 +80,7 @@ private:
    static std::uint32_t Swap(std::uint32_t val)
    {
 #if R__LITTLE_ENDIAN == 1
-      return RByteSwap<sizeof(fValBE)>::bswap(val);
+      return RByteSwap<sizeof(val)>::bswap(val);
 #else
       return val;
 #endif
@@ -104,7 +104,7 @@ private:
    static std::int32_t Swap(std::int32_t val)
    {
 #if R__LITTLE_ENDIAN == 1
-      return RByteSwap<sizeof(fValBE)>::bswap(val);
+      return RByteSwap<sizeof(val)>::bswap(val);
 #else
       return val;
 #endif
@@ -128,7 +128,7 @@ private:
    static std::uint64_t Swap(std::uint64_t val)
    {
 #if R__LITTLE_ENDIAN == 1
-      return RByteSwap<sizeof(fValBE)>::bswap(val);
+      return RByteSwap<sizeof(val)>::bswap(val);
 #else
       return val;
 #endif

--- a/tree/ntuple/v7/src/RMiniFile.cxx
+++ b/tree/ntuple/v7/src/RMiniFile.cxx
@@ -21,6 +21,7 @@
 #include <ROOT/RRawFile.hxx>
 #include <ROOT/RNTupleZip.hxx>
 
+#include <Byteswap.h>
 #include <TError.h>
 #include <TFile.h>
 #include <TKey.h>
@@ -35,6 +36,15 @@
 #include <string>
 #include <chrono>
 
+#ifndef R__LITTLE_ENDIAN
+#ifdef R__BYTESWAP
+// `R__BYTESWAP` is defined in RConfig.hxx for little-endian architectures; undefined otherwise
+#define R__LITTLE_ENDIAN 1
+#else
+#define R__LITTLE_ENDIAN 0
+#endif
+#endif /* R__LITTLE_ENDIAN */
+
 namespace {
 
 // The following types are used to read and write the TFile binary format
@@ -43,7 +53,14 @@ namespace {
 class RUInt16BE {
 private:
    std::uint16_t fValBE = 0;
-   static std::uint16_t Swap(std::uint16_t val) { return (val & 0x00FF) << 8 | (val & 0xFF00) >> 8; }
+   static std::uint16_t Swap(std::uint16_t val)
+   {
+#if R__LITTLE_ENDIAN == 1
+      return RByteSwap<sizeof(fValBE)>::bswap(val);
+#else
+      return val;
+#endif
+   }
 
 public:
    RUInt16BE() = default;
@@ -62,8 +79,11 @@ private:
    std::uint32_t fValBE = 0;
    static std::uint32_t Swap(std::uint32_t val)
    {
-      auto x = (val & 0x0000FFFF) << 16 | (val & 0xFFFF0000) >> 16;
-      return (x & 0x00FF00FF) << 8 | (x & 0xFF00FF00) >> 8;
+#if R__LITTLE_ENDIAN == 1
+      return RByteSwap<sizeof(fValBE)>::bswap(val);
+#else
+      return val;
+#endif
    }
 
 public:
@@ -83,8 +103,11 @@ private:
    std::int32_t fValBE = 0;
    static std::int32_t Swap(std::int32_t val)
    {
-      auto x = (val & 0x0000FFFF) << 16 | (val & 0xFFFF0000) >> 16;
-      return (x & 0x00FF00FF) << 8 | (x & 0xFF00FF00) >> 8;
+#if R__LITTLE_ENDIAN == 1
+      return RByteSwap<sizeof(fValBE)>::bswap(val);
+#else
+      return val;
+#endif
    }
 
 public:
@@ -104,9 +127,11 @@ private:
    std::uint64_t fValBE = 0;
    static std::uint64_t Swap(std::uint64_t val)
    {
-      auto x = (val & 0x00000000FFFFFFFF) << 32 | (val & 0xFFFFFFFF00000000) >> 32;
-      x = (x & 0x0000FFFF0000FFFF) << 16 | (x & 0xFFFF0000FFFF0000) >> 16;
-      return (x & 0x00FF00FF00FF00FF) << 8 | (x & 0xFF00FF00FF00FF00) >> 8;
+#if R__LITTLE_ENDIAN == 1
+      return RByteSwap<sizeof(fValBE)>::bswap(val);
+#else
+      return val;
+#endif
    }
 
 public:

--- a/tree/ntuple/v7/test/ntuple_endian.cxx
+++ b/tree/ntuple/v7/test/ntuple_endian.cxx
@@ -1,4 +1,5 @@
 // Override endianness detection in RColumnElement.hxx; assume big-endian machine
+// These tests are simulating a big endian machine; we will turn them off on an actual big endian node.
 #define R__LITTLE_ENDIAN 0
 
 #include "gtest/gtest.h"
@@ -104,6 +105,9 @@ public:
 
 TEST(RColumnElementEndian, ByteCopy)
 {
+#ifndef R__BYTESWAP
+   GTEST_SKIP() << "Skipping test on big endian node";
+#else
    ROOT::Experimental::Internal::RColumnElement<float, EColumnType::kReal32> element;
    EXPECT_EQ(element.IsMappable(), false);
 
@@ -121,10 +125,14 @@ TEST(RColumnElementEndian, ByteCopy)
    auto page2 = source1.PopulatePage(RPageStorage::ColumnHandle_t{}, NTupleSize_t{0});
    std::unique_ptr<unsigned char[]> buf2(static_cast<unsigned char *>(page2.GetBuffer())); // adopt buffer
    EXPECT_EQ(0, memcmp(buf2.get(), "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f", 16));
+#endif
 }
 
 TEST(RColumnElementEndian, Cast)
 {
+#ifndef R__BYTESWAP
+   GTEST_SKIP() << "Skipping test on big endian node";
+#else
    ROOT::Experimental::Internal::RColumnElement<std::int64_t, EColumnType::kInt32> element;
    EXPECT_EQ(element.IsMappable(), false);
 
@@ -146,10 +154,14 @@ TEST(RColumnElementEndian, Cast)
                        "\x00\x01\x02\x03\x00\x00\x00\x00\x04\x05\x06\x07\x00\x00\x00\x00"
                        "\x08\x09\x0a\x0b\x00\x00\x00\x00\x0c\x0d\x0e\x0f\x00\x00\x00\x00",
                        32));
+#endif
 }
 
 TEST(RColumnElementEndian, Split)
 {
+#ifndef R__BYTESWAP
+   GTEST_SKIP() << "Skipping test on big endian node";
+#else
    ROOT::Experimental::Internal::RColumnElement<double, EColumnType::kSplitReal64> splitElement;
 
    RPageSinkMock sink1(splitElement);
@@ -166,10 +178,14 @@ TEST(RColumnElementEndian, Split)
    auto page2 = source1.PopulatePage(RPageStorage::ColumnHandle_t{}, NTupleSize_t{0});
    std::unique_ptr<unsigned char[]> buf2(static_cast<unsigned char *>(page2.GetBuffer())); // adopt buffer
    EXPECT_EQ(0, memcmp(buf2.get(), "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f", 16));
+#endif
 }
 
 TEST(RColumnElementEndian, DeltaSplit)
 {
+#ifndef R__BYTESWAP
+   GTEST_SKIP() << "Skipping test on big endian node";
+#else
    using ClusterSize_t = ROOT::Experimental::ClusterSize_t;
 
    ROOT::Experimental::Internal::RColumnElement<ClusterSize_t, EColumnType::kSplitIndex32> element;
@@ -193,4 +209,5 @@ TEST(RColumnElementEndian, DeltaSplit)
                        "\x00\x01\x02\x03\x00\x00\x00\x00\x04\x05\x06\x07\x00\x00\x00\x00"
                        "\x08\x09\x0a\x0b\x00\x00\x00\x00\x0c\x0d\x0e\x0f\x00\x00\x00\x00",
                        32));
+#endif
 }

--- a/tree/ntuple/v7/test/ntuple_packing.cxx
+++ b/tree/ntuple/v7/test/ntuple_packing.cxx
@@ -97,11 +97,13 @@ TEST(Packing, HalfPrecisionFloat)
    element.Unpack(nullptr, nullptr, 0);
 
    float in = 3.14;
-   std::uint16_t b = 0;
-   element.Pack(&b, &in, 1);
-   EXPECT_EQ(0x4248, b); // Expected bit representation: 0b01000010 01001000
+   unsigned char buf[2] = {0, 0};
+   element.Pack(buf, &in, 1);
+   // Expected bit representation: 0b01000010 01001000
+   EXPECT_EQ(0x48, buf[0]);
+   EXPECT_EQ(0x42, buf[1]);
    float out = 0.;
-   element.Unpack(&out, &b, 1);
+   element.Unpack(&out, buf, 1);
    EXPECT_FLOAT_EQ(3.140625, out);
 
    float in4[] = {0.1, 0.2, 0.3, 0.4};
@@ -124,10 +126,10 @@ TEST(Packing, RColumnSwitch)
    element.Unpack(nullptr, nullptr, 0);
 
    ROOT::Experimental::RColumnSwitch s1(ClusterSize_t{0xaa}, 0x55);
-   std::pair<std::uint64_t, std::uint32_t> out;
-   element.Pack(&out, &s1, 1);
+   unsigned char out[12];
+   element.Pack(out, &s1, 1);
    ROOT::Experimental::RColumnSwitch s2;
-   element.Unpack(&s2, &out, 1);
+   element.Unpack(&s2, out, 1);
    EXPECT_EQ(0xaa, s2.GetIndex());
    EXPECT_EQ(0x55, s2.GetTag());
 }

--- a/tree/ntuple/v7/test/ntuple_serialize.cxx
+++ b/tree/ntuple/v7/test/ntuple_serialize.cxx
@@ -1,5 +1,7 @@
 #include "ntuple_test.hxx"
 
+#include <Byteswap.h>
+
 TEST(RNTuple, SerializeInt)
 {
    std::int32_t value;
@@ -113,6 +115,10 @@ TEST(RNTuple, SerializeEnvelope)
       std::uint64_t payload = 0;
       std::uint64_t xxhash3 = 0;
    } testEnvelope;
+#ifndef R__BYTESWAP
+   // big endian
+   testEnvelope.typeAndSize = RByteSwap<8>::bswap(testEnvelope.typeAndSize);
+#endif
 
    EXPECT_EQ(8u, RNTupleSerializer::SerializeEnvelopePostscript(reinterpret_cast<unsigned char *>(&testEnvelope), 16));
    testEnvelope.xxhash3 = 0;
@@ -122,7 +128,12 @@ TEST(RNTuple, SerializeEnvelope)
    } catch (const RException& err) {
       EXPECT_THAT(err.what(), testing::HasSubstr("XxHash-3"));
    }
+#ifndef R__BYTESWAP
+   // big endian
+   testEnvelope.typeAndSize = RByteSwap<8>::bswap(137);
+#else
    testEnvelope.typeAndSize = 137;
+#endif
 
    EXPECT_EQ(8u, RNTupleSerializer::SerializeEnvelopePostscript(reinterpret_cast<unsigned char *>(&testEnvelope), 16));
    try {
@@ -131,7 +142,12 @@ TEST(RNTuple, SerializeEnvelope)
    } catch (const RException &err) {
       EXPECT_THAT(err.what(), testing::HasSubstr("envelope type mismatch"));
    }
+#ifndef R__BYTESWAP
+   // big endian
+   testEnvelope.typeAndSize = RByteSwap<8>::bswap(137);
+#else
    testEnvelope.typeAndSize = 137;
+#endif
 
    EXPECT_EQ(8u, RNTupleSerializer::SerializeEnvelopePostscript(reinterpret_cast<unsigned char *>(&testEnvelope), 16));
    try {
@@ -141,7 +157,12 @@ TEST(RNTuple, SerializeEnvelope)
       EXPECT_THAT(err.what(), testing::HasSubstr("envelope buffer size too small"));
    }
 
+#ifndef R__BYTESWAP
+   // big endian
+   testEnvelope.typeAndSize -= uint64_t(9) << 40;
+#else
    testEnvelope.typeAndSize -= 9 << 16;
+#endif
    try {
       RNTupleSerializer::DeserializeEnvelope(&testEnvelope, sizeof(testEnvelope), 137).Unwrap();
       FAIL() << "too small envelope buffer should throw";
@@ -149,7 +170,12 @@ TEST(RNTuple, SerializeEnvelope)
       EXPECT_THAT(err.what(), testing::HasSubstr("invalid envelope, too short"));
    }
 
+#ifndef R__BYTESWAP
+   // big endian
+   testEnvelope.typeAndSize = RByteSwap<8>::bswap(137);
+#else
    testEnvelope.typeAndSize = 137;
+#endif
    std::uint64_t xxhash3_write;
    RNTupleSerializer::SerializeEnvelopePostscript(reinterpret_cast<unsigned char *>(&testEnvelope), 16, xxhash3_write);
    std::uint64_t xxhash3_read;
@@ -352,7 +378,12 @@ TEST(RNTuple, SerializeLocator)
    EXPECT_EQ(1337U, locator.GetPosition<RNTupleLocatorObject64>().fLocation);
 
    std::int32_t *head = reinterpret_cast<std::int32_t *>(buffer);
+#ifndef R__BYTESWAP
+   // on big endian system
+   *head = *head | 0x3;
+#else
    *head = (0x3 << 24) | *head;
+#endif
    try {
       RNTupleSerializer::DeserializeLocator(buffer, 16, locator).Unwrap();
       FAIL() << "unsupported locator type should throw";

